README.md

```diff
@@ -1,2 +1,1584 @@
+networking and data requirements.
+
+This masterclass acts as an introduction to how both single and multicontainer
+applications operate, as well as demonstrating some of their most common uses
+and some advanced techniques.
+
+# Hardware and Software Requirements
+
+It is assumed that the reader has access to the following:
+
+* A balena supported device, such as a [balenaFin 1.1](https://store.balena.io/collections/developer-kit/products/balenafin-v1-1-0-developer-kit),
+        [Raspberry Pi 3](https://www.raspberrypi.org/products/raspberry-pi-3-model-b/)
+        or [Intel NUC](https://www.intel.co.uk/content/www/uk/en/products/boards-kits/nuc.html). If you don't have a device, you can emulate an Intel NUC by
+        installing VirtualBox and following [this guide](https://www.balena.io/blog/no-hardware-use-virtualbox/)
+* A suitable text editor for developing code on your development platform (eg.
+    [Visual Code](https://code.visualstudio.com/))
+* A suitable shell environment for command execution (such as `bash`)
+* A [balenaCloud](https://www.balena.io/) account
+* Some familiarity with [Dockerfiles](https://docs.docker.com/engine/reference/builder/) and [docker-compose]()
+* An installation of [balena CLI](https://github.com/balena-io/balena-cli/blob/master/INSTALL.md)
+
+# Exercises
+
+All of the following exercises assume that you are able to access a suitable
+Unix based shell and that balena CLI is installed. The exercises include
+commands which can be run in such a shell, and are represented by a line
+prefixed with `$`. Information returned from execution of a command may be
+appended under the line to show what might be returned. For example:
+
+```
+$ balena version
+11.11.2
+```
+
+# 1. Service Images, Containers, Volumes and Networks
+
+balena applications are comprised of service images, containers and volumes.
+When an application is built, one or more images are generated by the builders
+and then stored in balena's registries. When a device downloads an application,
+it retrieves these images from the registries and then stores them locally.
+
+The device Supervisor then looks at the images (and any docker-compose manifest
+associated with them) to determine how to create those services as containers,
+which form executable instances of an image.
+
+Additionally, volumes (persistent data stores) can be bound to an executble
+container to ensure that data exists across instances of the images.
+
+## 1.1 Service Images
+
+Service images are the basis for any service. They comprise the binaries and
+data for running services, but they also contain metadata that informs
+balenaEngine on how to initialise and configure the service. As such, an image
+is essentially a self-enclosed Linux filesystem with everything required for
+running an instance of Linux, except for the kernel.
+
+Images are produced by the balena builders, which use Dockerfiles to determine
```

Comment:
they could of course be built anywhere else (localmode or just as `image:` reference in a compose. but i think its probably best not to get to into the weeds here.

Reply:
Yeah, have clarified potential sources.